### PR TITLE
[RAPTOR-6035]Perform schema checks for transform output

### PR DESCRIPTION
**Pull request by scottp-ml**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Apply schema checks when running transform tasks.

### CODE REVIEW

File: custom_model_runner/CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,13 @@ All notable changes to this project will be documented in this file.
...
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-#### [1.5.10] - Current
+#### [1.5.11] - Current
+##### Fixed
+- Apply default schema to transforms
+##### Added 
+- type schema to pipeline examples
+
+#### [1.5.10] - 2021-08-13
```
Comment (jmbannon) [inline custom_model_runner/CHANGELOG.md:13]: This was release 2021-07-30, can you update that

File: tests/drum/test_fit.py [hunk 2]
```diff
@@ -275,11 +275,9 @@ def test_fit_hyperparameters(
             (SKLEARN_TRANSFORM, PYTHON),
             (SKLEARN_TRANSFORM_WITH_Y, PYTHON),
             (SKLEARN_TRANSFORM_NO_HOOK, PYTHON),
-            (SKLEARN_TRANSFORM_NON_NUMERIC, PYTHON),
-            (R_TRANSFORM, R_FIT),
-            (R_TRANSFORM_NO_Y, R_FIT),
-            (R_TRANSFORM_NO_HOOK, R_FIT),
-            (R_TRANSFORM_NON_NUMERIC, R_FIT),
+            # (R_TRANSFORM, R_FIT),
```
Comment (jmbannon) [inline tests/drum/test_fit.py:278]: uncomment

File: tests/drum/test_fit.py [hunk 3]
```diff
@@ -662,6 +710,7 @@ def test_fit_schema_failure(
 
         # The predict server will not return the full stacktrace since it is ran in a forked process
         if error_in_predict_server:
-            assert "ERROR: schema validation failed for output" in stdout
+            print(stdout)
```
Comment (jmbannon) [inline tests/drum/test_fit.py:713]: remove

File: tests/drum/test_fit.py [hunk 4]
```diff
@@ -319,6 +317,56 @@ def test_transform_fit(
...
+            cmd = _cmd_add_class_labels(
+                cmd, resources.class_labels(framework, problem), target_type=target_type
+            )
+
+        cmd += weights_cmd
+
+        if strict:
+            with pytest.raises(AssertionError):
+                _, stdout, _ = _exec_shell_cmd(
+                    cmd, "Failed in {} command line! {}".format(ArgumentsOptions.MAIN_COMMAND, cmd)
```
Comment (jmbannon) [inline tests/drum/test_fit.py:363]: Checking for the actual validation error string here would be idea. Is this what jenkins was giving you trouble with?

File: public_dropin_environments/java_codegen/env_info.json [hunk 5]
```diff
@@ -3,6 +3,6 @@
   "name": "[DataRobot] Java Drop-In (DR Codegen, H2O)",
   "description": "This template can be used as an environment for DataRobot generated scoring code or models that implement the either the IClassificationPredictor or IRegressionPredictor interface from the datarobot-prediction package and for H2O models exported as POJO or MOJO.",
   "programmingLanguage": "java",
-  "environmentVersionId": "610305345da4263681a176b0",
+  "environmentVersionId": "6116a8ac90cff07c9f4cb91f",
```
Comment (jmbannon) [inline public_dropin_environments/java_codegen/env_info.json:6]: These should not be updated until we do a pypi release

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 6]
```diff
@@ -187,6 +187,26 @@ def _get_mapping(field: Fields, values: List[Values]) -> Map:
...
+        return True
+    else:
+        return False
+
+
+def check_csr_sparse(data):
+    try:
+        return data.format == "csr"
+    except AttributeError:
+        return data.dtypes.apply(pd.api.types.is_sparse).any()
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:201]: Where and why do we load sparse data into a pandas dataframe?

File: tests/drum/test_fit.py (hunk 4 above)
Comment (scottp-ml) [reply tests/drum/test_fit.py:363]: no, this mostly matches the test above for the cases that pass, but with the pytest.raises.

File: tests/drum/test_fit.py [hunk 7]
```diff
@@ -662,6 +710,7 @@ def test_fit_schema_failure(
 
         # The predict server will not return the full stacktrace since it is ran in a forked process
         if error_in_predict_server:
-            assert "ERROR: schema validation failed for output" in stdout
+            print(stdout)
+            assert "expected types to exactly match: CAT" in stdout
```
Comment (scottp-ml) [inline tests/drum/test_fit.py:714]: @jmbannon this is the one that was causing issues

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 6 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:201]: if the transform returns it as sparse, when it is loaded to test in the type schema.

Comment (jmbannon) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:201]: But where do we put a CSR matrix inside of a pandas dataframe? I feel like we should just be passing the CSR itself

**Review by jmbannon — APPROVED** [review]

File: custom_model_runner/datarobot_drum/drum/perf_testing.py [hunk 8]
```diff
@@ -674,6 +676,10 @@ def check_transform_server(self, target_temp_location=None):
                 payload.update({"y": open(target_location)})
 
             response = requests.post(run.url_server_address + endpoint, files=payload)
+            transformed_values = read_x_data_from_response(response)
+            if self._schema_validator:
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/perf_testing.py:680]: when would the validator not be present?

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 9]
```diff
@@ -187,6 +187,26 @@ def _get_mapping(field: Fields, values: List[Values]) -> Map:
     return Map({"field": Enum(str(field)), "condition": conditions, "value": value_enum})
 
 
+def check_is_mm(data):
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:190]: why do we need this? MM data should be loaded in as a sparse DF.

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 10]
```diff
@@ -187,6 +187,26 @@ def _get_mapping(field: Fields, values: List[Values]) -> Map:
...
+def check_csr_sparse(data):
+    try:
+        return data.format == "csr"
+    except AttributeError:
+        return False
+
+
+def check_is_sparse(data):
+    # this is more verbose than ideal since or doesn't seem to correctly short circuit with the data series apply,
+    # causing an issue dtypes missing in this case.
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:206]: this comment is hard to understand

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 11]
```diff
@@ -278,18 +298,27 @@ def validate(self, dataframe):
...
+            # only numeric can be a csr or matrix market sparse matrix
+            types[Values.NUM] = True
+            types[Values.TXT] = False
+            types[Values.IMG] = False
+            types[Values.CAT] = False
+            types[Values.DATE] = False
+        else:
+            types[Values.NUM] = dataframe.select_dtypes(np.number).shape[1] > 0
+            txt_columns = self.number_of_text_columns(dataframe)
+            img_columns = self.number_of_img_columns(dataframe)
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:312]: maybe add an `> 0` to the end of these?

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 12]
```diff
@@ -278,18 +298,27 @@ def validate(self, dataframe):
...
+            types[Values.TXT] = txt_columns > 0
+            types[Values.IMG] = img_columns > 0
+            types[Values.CAT] = (
+                dataframe.select_dtypes("O").shape[1]
+                - (txt_columns + img_columns)
+                + dataframe.select_dtypes("boolean").shape[1]
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:318]: could you clean this up somehow? it looks a little messy as is. 

maybe create variables for num_obj_columns and num_boolean_columns.

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 13]
```diff
@@ -320,7 +349,7 @@ def __init__(self, condition, values):
 
     def validate(self, dataframe):
 
-        is_sparse = dataframe.dtypes.apply(pd.api.types.is_sparse).any()
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:323]: so why was this a bad technique?

File: custom_model_runner/datarobot_drum/resource/transform_helpers.py [hunk 14]
```diff
@@ -93,3 +93,9 @@ def parse_multi_part_response(response):
         parsed_response.update({key: value})
 
     return parsed_response
+
+
+def read_x_data_from_response(response):
+    reader = {"arrow": read_arrow_payload, "sparse": read_mtx_payload, "csv": read_csv_payload}
+    data = parse_multi_part_response(response)
+    return reader[data[X_FORMAT_KEY]](data, X_TRANSFORM_KEY)
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/resource/transform_helpers.py:101]: this is a cool structure!

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 13 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:323]: it catches if there is a sparse series in the dataframe, not if the dataframe itself is sparse

File: tests/drum/unit/test_typeschema_validation.py [hunk 15]
```diff
@@ -140,7 +140,7 @@ def missing_data(self, data):
 
     @pytest.fixture
     def sparse_df(self):
-        yield pd.DataFrame.sparse.from_spmatrix(sparse.eye(10))
+        yield sparse.csr_matrix(sparse.eye(10))
```
Comment (4dahalibut) [inline tests/drum/unit/test_typeschema_validation.py:143]: Why was this change necessary?

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 13 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:323]: what does it mean for the dataframe to be sparse, if not that it has sparse series?

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 11 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:312]: these are suppose to be counts, not bools

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 13 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:323]: like a csr dataframe object, which is what is actually received

File: tests/drum/unit/test_typeschema_validation.py (hunk 15 above)
Comment (scottp-ml) [reply tests/drum/unit/test_typeschema_validation.py:143]: to correctly match what is received when sparse data is returned from a transform.

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 9 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:190]: this is how it was read from the transform in some of the functional tests.

File: custom_model_runner/datarobot_drum/drum/perf_testing.py (hunk 8 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/perf_testing.py:680]: if the run mode isn't fit or push.

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 10 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:206]: old comment that is no longer relevant, removing

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 13 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:323]: do you mean a csr matrix?

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 9 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:190]: lets please assert that we're passing around a dataframe with sparse columns, and not a csr matrix, if possible.

File: custom_model_runner/datarobot_drum/drum/perf_testing.py (hunk 8 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/perf_testing.py:680]: this is only called for fit. I don't think you need the check.

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 13 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:323]: yes

File: custom_model_runner/datarobot_drum/drum/perf_testing.py (hunk 8 above)
Comment (scottp-ml) [reply custom_model_runner/datarobot_drum/drum/perf_testing.py:680]: ok

File: custom_model_runner/datarobot_drum/drum/perf_testing.py [hunk 16]
```diff
@@ -674,6 +676,9 @@ def check_transform_server(self, target_temp_location=None):
                 payload.update({"y": open(target_location)})
 
             response = requests.post(run.url_server_address + endpoint, files=payload)
+            transformed_values = read_x_data_from_response(response)
+            # Validate that the transformed values are of the type and shape the user specified in the schema
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/perf_testing.py:680]: this comment should be in the below function

File: custom_model_runner/datarobot_drum/drum/perf_testing.py [hunk 17]
```diff
@@ -744,9 +749,8 @@ def check_prediction_side_effects(self):
 
             preds_full_subset = preds_full.iloc[data_subset.index]
 
-            if self._schema_validator:
-                # Validate that the predictions are of the type and shape the user specified in the schema
-                self._schema_validator.validate_outputs(preds_sample)
+            # Validate that the predictions are of the type and shape the user specified in the schema
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/perf_testing.py:752]: this comment should be in the below function

File: custom_model_runner/datarobot_drum/resource/transform_helpers.py [hunk 18]
```diff
@@ -93,3 +93,13 @@ def parse_multi_part_response(response):
         parsed_response.update({key: value})
 
     return parsed_response
+
+
+def read_x_data_from_response(response):
+    reader = {
+        "arrow": read_arrow_payload,
+        "sparse": lambda d, k: pd.DataFrame.sparse.from_spmatrix(read_mtx_payload(d, k)),
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/resource/transform_helpers.py:101]: I'd rather see an internal function instead of a lambda. It's more readable that way, but this is fine.

**Review by 4dahalibut — APPROVED** [review]
